### PR TITLE
798 implement dataobjectstateincrementalmode for deltalake and icebergtabledataobject

### LINKED ISSUE #798: Implement DataObjectStateIncrementalMode for DeltaLake- and IcebergTableDataObject
**Is your feature request related to a problem? Please describe.**
Load data incrementally from DeltaLake and Iceberg Tables is done most efficiently using their CDC data stream:
https://docs.delta.io/latest/delta-change-data-feed.html#enable-change-data-feed&language-scala
https://www.databricks.com/blog/2021/06/09/how-to-simplify-cdc-with-delta-lakes-change-data-feed.html

SDLB supports reading incrementally from DataObjects with separate state using DataObjectStateIncrementalMode. The input DataObject is responsible to create state for remembering the records read in the last run and must implement trait CanCreateIncrementalOutput for this. DeltaLake- and IcebergTableDataObject dont support this yet.

**Describe the solution you'd like**
Implement trait CanCreateIncrementalOutput for DeltaLake- and IcebergTableDataObject. They should use version as state.

**Additional context**
To be discussed what we do with the additional CDC column _change_type. For the first implementation just the original table columns should be in the DataFrame created, and _change_type should be filtered on "insert" and "update_postimage".

For the future we could think of creating an additional trait CanCreateCDCIncrementalOutput, which would include standardized columns for CDC-operation. An Action like Deduplicate- or HistorizeAction could switch the input DataObject in CDC Mode if supported, and use the CDC-operation column as input for the merge statement.

### PR DESCRIPTION
**Pull request by ThierryGirod**
### What changes are included in the pull request?
 
Updates of IcebergTableDataObject and DeltaLakeTableDataObject class to support incremental load. Implemented trait CanCreateIncrementalOutput on both classes for state handling. Added test cases. 

### Why are the changes needed?
Implement issue #798

### CODE REVIEW

File: sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala [hunk 1]
```diff
@@ -199,17 +200,65 @@ case class DeltaLakeTableDataObject(override val id: DataObjectId,
...
+  }
+
+  private def propertyExistsWithValue(name: String, value: String) (implicit session: SparkSession): Boolean = {
+    val details = DeltaTable.forName(session, table.fullName).detail()
+    val properties = details.select("properties").head.getMap[String, String](0)
+
+    properties.exists(_ == name -> value)
+  }
+
+  @transient private val EnableCdcFeedProperty = "delta.enableChangeDataFeed"
```
Comment (zzeekk) [inline sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala]: Variable names should start with lowercase:
```suggestion
  @transient private val enableCdcFeedProperty = "delta.enableChangeDataFeed"
```

File: sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala [hunk 2]
```diff
@@ -199,17 +200,65 @@ case class DeltaLakeTableDataObject(override val id: DataObjectId,
...
   override def getSparkDataFrame(partitionValues: Seq[PartitionValues] = Seq())(implicit context: ActionPipelineContext): DataFrame = {
-    val df = context.sparkSession.table(table.fullName)
+
+    implicit val session: SparkSession = context.sparkSession
+
+    val cdcActivated = propertyExistsWithValue(EnableCdcFeedProperty, "true")
+
+    val df = if(cdcActivated && incrementalOutputExpr.isDefined) {
+
+      require(table.primaryKey.isDefined, s"PrimaryKey for table [${table.fullName}] needs to be defined when using DataObjectStateIncrementalMode")
```
Comment (zzeekk) [inline sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala]:
```suggestion
      require(table.primaryKey.isDefined, s"($id) PrimaryKey for table [${table.fullName}] needs to be defined when using DataObjectStateIncrementalMode")
```

File: sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala [hunk 3]
```diff
@@ -199,17 +200,65 @@ case class DeltaLakeTableDataObject(override val id: DataObjectId,
...
+
+      val windowSpec = Window.partitionBy(table.primaryKey.get.mkString(",")).orderBy(col("_commit_timestamp").desc)
+
+      context.sparkSession.read.format("delta")
+        .option("readChangeFeed", "true")
+        .option("startingVersion", incrementalOutputExpr.get)
+        .table(table.fullName)
+        .where(expr("_change_type IN ('insert','update_postimage')"))
+        .withColumn("rank", rank().over(windowSpec))
```
Comment (zzeekk) [inline sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala]: What if the DataFrame already has a "rank" column?
I would prefix this with "_" as we do with most helper columns, e.g. "_rank"

File: sdl-deltalake/src/test/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObjectTest.scala [hunk 4]
```diff
@@ -320,4 +323,144 @@ class DeltaLakeTableDataObjectTest extends FunSuite with BeforeAndAfter {
...
+
+    // test
+    val thrown = intercept[IllegalArgumentException] {
+      targetDO.setState(newState2)
+      targetDO.getSparkDataFrame()(contextExec).count()
+    }
+
+    // check
+    assert(thrown.isInstanceOf[IllegalArgumentException])
+    assert(thrown.getMessage == s"requirement failed: PrimaryKey for table [${targetDO.table.fullName}] needs to be defined when using DataObjectStateIncrementalMode")
```
Comment (zzeekk) [inline sdl-deltalake/src/test/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObjectTest.scala]: I would simplify the check (i just added $(id) to the message in a previous comment...)
```suggestion
    assert(thrown.getMessage.contains("PrimaryKey for table"))
```

File: sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala [hunk 5]
```diff
@@ -199,17 +200,65 @@ case class DeltaLakeTableDataObject(override val id: DataObjectId,
...
+
+    implicit val session: SparkSession = context.sparkSession
+
+    val cdcActivated = propertyExistsWithValue(EnableCdcFeedProperty, "true")
+
+    val df = if(cdcActivated && incrementalOutputExpr.isDefined) {
+
+      require(table.primaryKey.isDefined, s"PrimaryKey for table [${table.fullName}] needs to be defined when using DataObjectStateIncrementalMode")
+
+      val windowSpec = Window.partitionBy(table.primaryKey.get.mkString(",")).orderBy(col("_commit_timestamp").desc)
```
Comment (zzeekk) [inline sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala]: No need for mkString:
```suggestion
      val windowSpec = Window.partitionBy(table.primaryKey.get.map(col): _*).orderBy(col("_commit_timestamp").desc)
```

File: sdl-iceberg/src/main/scala/io/smartdatalake/workflow/dataobject/IcebergTableDataObject.scala [hunk 6]
```diff
@@ -225,9 +227,40 @@ case class IcebergTableDataObject(override val id: DataObjectId,
   }
 
   override def getSparkDataFrame(partitionValues: Seq[PartitionValues] = Seq())(implicit context: ActionPipelineContext): DataFrame = {
-    val df = context.sparkSession.table(table.fullName)
+
+    val df = incrementalOutputExpr match {
+      case Some(snapshotId) =>
+        require(table.primaryKey.isDefined, s"PrimaryKey for table [${table.fullName}] needs to be defined when using DataObjectStateIncrementalMode")
```
Comment (zzeekk) [inline sdl-iceberg/src/main/scala/io/smartdatalake/workflow/dataobject/IcebergTableDataObject.scala]:
```suggestion
        require(table.primaryKey.isDefined, s"($id) PrimaryKey for table [${table.fullName}] needs to be defined when using DataObjectStateIncrementalMode")
```

File: sdl-iceberg/src/main/scala/io/smartdatalake/workflow/dataobject/IcebergTableDataObject.scala [hunk 7]
```diff
@@ -225,9 +227,40 @@ case class IcebergTableDataObject(override val id: DataObjectId,
...
+        val icebergTable = if(snapshotId.equalsIgnoreCase("0")) context.sparkSession.table(table.fullName)
```
Comment (zzeekk) [inline sdl-iceberg/src/main/scala/io/smartdatalake/workflow/dataobject/IcebergTableDataObject.scala]: In Scala this is better readable:
```suggestion
        val icebergTable = if(snapshotId == "0") context.sparkSession.table(table.fullName)
```

File: sdl-iceberg/src/main/scala/io/smartdatalake/workflow/dataobject/IcebergTableDataObject.scala [hunk 8]
```diff
@@ -225,9 +227,40 @@ case class IcebergTableDataObject(override val id: DataObjectId,
...
+            s"""CALL ${getIcebergCatalog.name}.system.create_changelog_view(table => '${getIdentifier.toString}'
+               |, options => map('start-snapshot-id', '${snapshotId}')
+               |, compute_updates => true
+               |, identifier_columns => array('${table.primaryKey.get.mkString("','")}')
+               |)""".stripMargin)
+
+          // read cdc events
+          val temporaryViewName = table.name + "_changes"
+
+          val windowSpec = Window.partitionBy(table.primaryKey.get.mkString(",")).orderBy(col("_change_ordinal").desc)
```
Comment (zzeekk) [inline sdl-iceberg/src/main/scala/io/smartdatalake/workflow/dataobject/IcebergTableDataObject.scala]:
```suggestion
          val windowSpec = Window.partitionBy(table.primaryKey.get.map(col): _*).orderBy(col("_change_ordinal").desc)
```

File: sdl-iceberg/src/test/scala/io/smartdatalake/workflow/dataobject/IcebergTableDataObjectTest.scala [hunk 9]
```diff
@@ -295,4 +295,112 @@ class IcebergTableDataObjectTest extends FunSuite with BeforeAndAfter {
...
+
+    // test
+    val thrown = intercept[IllegalArgumentException] {
+      targetDO.setState(newState1)
+      targetDO.getSparkDataFrame()(contextExec).count()
+    }
+
+    // check
+    assert(thrown.isInstanceOf[IllegalArgumentException])
+    assert(thrown.getMessage == s"requirement failed: PrimaryKey for table [${targetDO.table.fullName}] needs to be defined when using DataObjectStateIncrementalMode")
```
Comment (zzeekk) [inline sdl-iceberg/src/test/scala/io/smartdatalake/workflow/dataobject/IcebergTableDataObjectTest.scala]:
```suggestion
    assert(thrown.getMessage.contains(s"PrimaryKey for table"))
```

**Review by zzeekk — APPROVED** [review]
Looks good!
Some comments to fix, but otherwise good to merge.

File: sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala (hunk 3 above)
Comment (ThierryGirod) [reply sdl-deltalake/src/main/scala/io/smartdatalake/workflow/dataobject/DeltaLakeTableDataObject.scala]: Fixed for both DeltaLakeTableDataObject and IcebergTableDataObject in 2aa23d7e0ba1977d76e85c551d068e5552978d7e

**Pull request merged**